python/ray/dag/dag_node.py

```diff
@@ -335,6 +354,33 @@ def __call__(self, node: "DAGNode"):
             )
         )
 
+    def bfs(self, fn: "Callable[[DAGNode], T]"):
+        """Breadth-first search on the DAG."""
+        visited = set()
+        queue = [self]
+        while queue:
+            node = queue.pop(0)
+            if node not in visited:
+                fn(node)
+                visited.add(node)
+                # Add all unseen downstream and upstream nodes to the queue.
+                # This function should be called by the root of the DAG. However,
+                # in some invalid cases, some nodes may not be descendants of the
```

Comment:
Is this valid DAG? If it's not even valid DAG, we should throw an error at DAG layer rather than aDAG (compilation) layer.

Reply:
> Is this valid DAG?

Not sure

> If it's not even valid DAG, we should throw an error at DAG layer rather than aDAG (compilation) layer.

It makes sense to me. I think we should unify the codepath like

* Step 1: Build a DAG.
* Step 2: Validate the DAG (e.g., check whether any nodes are not descendants of InputNode).
* Step 3:
  * Option 1: Execute (DAG)
  * Option 2: Compile -> Execute (ADAG)


How about opening an issue to track the progress?

1. This PR is compatible with the current behavior, where we validate these edge cases during aDAG compilation instead of building graphs.

2. Currently, `bfs` is only used by ADAG. It doesn't affect DAG.

Reply:
IIC replacing apply_recursive will solve the issue?

Reply:
`apply_recursive` currently doesn't validate the case where not all DAG nodes are descendants of `InputNode`. This validation is currently handled by ADAG. For example, `a2.inc.bind(1)` is not a descendant of `InputNode`.

https://github.com/ray-project/ray/blob/6743d3d6efe16c165edbac7c1556095168bb51f1/python/ray/dag/tests/experimental/test_accelerated_dag.py#L661-L669

Reply:
> in some invalid cases

BTW, do we know what are those invalid cases? Leaf node cases can be handled here if we traverse from the root, right? So these are some other cases?

Would be nice if we could be more specific in the comment.

Reply:
> apply_recursive currently doesn't validate the case where not all DAG nodes are descendants of InputNode. This validation is currently handled by ADAG. For example, a2.inc.bind(1) is not a descendant of InputNode.

ah I meant if we make regular dag use bfs instead of apply_recursive, this validation should be done the same way for dag & adag?

Reply:
> BTW, do we know what are those invalid cases? Leaf node cases can be handled here if we traverse from the root, right? So these are some other cases?

Update in the comment


> ah I meant if we make regular dag use bfs instead of apply_recursive, this validation should be done the same way for dag & adag?

Yes